File: ports/java/src/main/java/elevatorsaga/World.java
package elevatorsaga;

import java.util.*;

/**
 * The game world containing elevators, floors, and users.
 */
public class World extends Observable {
    public final double floorHeight;
    public int transportedCounter = 0;
    public double transportedPerSec = 0.0;
    public int moveCount = 0;
    public double elapsedTime = 0.0;
    public double maxWaitTime = 0.0;
    public double avgWaitTime = 0.0;
    public boolean challengeEnded = false;
    public List<User> users = new ArrayList<>();
    public List<Floor> floors;
    public List<Elevator> elevators;
    public List<ElevatorInterface> elevatorInterfaces;

    private final double spawnRate;
    private double elapsedSinceSpawn;
    private final int floorCount;
    private final Random random = new Random();

    public World(Map<String, Object> options) {
        Map<String, Object> defaults = new HashMap<>();
        defaults.put("floor_height", 50.0);
        defaults.put("floor_count", 4);
        defaults.put("elevator_count", 2);
        defaults.put("spawn_rate", 0.5);
        defaults.put("elevator_capacities", new int[]{4});

        if (options != null) defaults.putAll(options);

        this.floorHeight = ((Number) defaults.get("floor_height")).doubleValue();
        this.floorCount = ((Number) defaults.get("floor_count")).intValue();
        int elevatorCount = ((Number) defaults.get("elevator_count")).intValue();
        this.spawnRate = ((Number) defaults.get("spawn_rate")).doubleValue();
        int[] elevatorCapacities = (int[]) defaults.get("elevator_capacities");
        this.elapsedSinceSpawn = 1.001 / spawnRate;

        Floor.ErrorHandler errorHandler = e -> trigger("usercode_error", e);

        // Create floors
        floors = new ArrayList<>();
        for (int i = 0; i < floorCount; i++) {
            double yPos = (floorCount - 1 - i) * floorHeight;
            floors.add(new Floor(i, yPos, errorHandler));
        }

        // Create elevators
        elevators = new ArrayList<>();
        double currentX = 200.0;
        for (int i = 0; i < elevatorCount; i++) {
            int cap = elevatorCapacities[i % elevatorCapacities.length];
            Elevator elevator = new Elevator(2.6, floorCount, floorHeight, cap);
            elevator.moveTo(currentX, null);
            elevator.setFloorPosition(0);
            elevator.updateDisplayPosition();
            currentX += 20 + elevator.width;
            elevators.add(elevator);
        }

        // Create interfaces
        elevatorInterfaces = new ArrayList<>();
        for (Elevator e : elevators) {
            elevatorInterfaces.add(new ElevatorInterface(e, floorCount, errorHandler));
        }

        // Bind elevator entrance events
        for (Elevator elevator : elevators) {
            elevator.on("entrance_available", args -> handleElevAvailability((Elevator) args[0]));
        }

        // Bind floor button repressing
        for (Floor floor : floors) {
            floor.on("up_button_pressed", args -> handleButtonRepressing("up_button_pressed", (Floor) args[0]));
            floor.on("down_button_pressed", args -> handleButtonRepressing("down_button_pressed", (Floor) args[0]));
        }
    }

    private void handleElevAvailability(Elevator elevator) {
        for (int i = 0; i < floors.size(); i++) {
            if (elevator.currentFloor == i) {
                floors.get(i).elevatorAvailable(elevator);
            }
        }
        for (User user : new ArrayList<>(users)) {
            if (user.currentFloor == elevator.currentFloor) {
                user.elevatorAvailable(elevator, floors.get(elevator.currentFloor));
            }
        }
    }

    private void handleButtonRepressing(String eventName, Floor floor) {
        int offset = random.nextInt(elevators.size());
        for (int i = 0; i < elevators.size(); i++) {
            int elevIndex = (i + offset) % elevators.size();
            Elevator elevator = elevators.get(elevIndex);
            if ((eventName.equals("up_button_pressed") && elevator.goingUpIndicator) ||
                (eventName.equals("down_button_pressed") && elevator.goingDownIndicator)) {
                if (elevator.currentFloor == floor.level && elevator.isOnAFloor() &&
                    !elevator.isMoving && !elevator.isFull()) {
                    elevatorInterfaces.get(elevIndex).goToFloor(floor.level, true);
                    return;
                }
            }
        }
    }

    private void recalculateStats() {
        if (elapsedTime > 0) {
            transportedPerSec = transportedCounter / elapsedTime;
        }
        moveCount = 0;
        for (Elevator e : elevators) moveCount += e.moveCount;
        trigger("stats_changed");
    }

    private void registerUser(User user) {
        users.add(user);
        user.updateDisplayPosition(true);
        user.spawnTimestamp = elapsedTime;
        trigger("new_user", user);

        user.on("exited_elevator", args -> {
            transportedCounter++;
            maxWaitTime = Math.max(maxWaitTime, elapsedTime - user.spawnTimestamp);
            avgWaitTime = (avgWaitTime * (transportedCounter - 1) + (elapsedTime - user.spawnTimestamp)) / transportedCounter;
            recalculateStats();
        });
        user.updateDisplayPosition(true);
    }

    private User spawnUserRandomly() {
        double weight = 55 + random.nextInt(46);
        User user = new User(weight);
        if (random.nextInt(41) == 0) {
            user.displayType = "child";
        } else if (random.nextInt(2) == 0) {
            user.displayType = "female";
        } else {
            user.displayType = "male";
        }

        user.moveTo(105.0 + random.nextInt(41), null);
        int currentFloor = random.nextInt(2) == 0 ? 0 : random.nextInt(floorCount);
        int destinationFloor;
        if (currentFloor == 0) {
            destinationFloor = 1 + random.nextInt(floorCount - 1);
        } else {
            if (random.nextInt(11) == 0) {
                destinationFloor = (currentFloor + 1 + random.nextInt(floorCount - 1)) % floorCount;
            } else {
                destinationFloor = 0;
            }
        }
        user.appearOnFloor(floors.get(currentFloor), destinationFloor);
        return user;
    }

    public void update(double dt) {
        elapsedTime += dt;
        elapsedSinceSpawn += dt;

        while (elapsedSinceSpawn > 1.0 / spawnRate) {
            elapsedSinceSpawn -= 1.0 / spawnRate;
            registerUser(spawnUserRandomly());
        }

        for (Elevator e : elevators) {
            e.update(dt);
            e.updateElevatorMovement(dt);
        }

        for (User u : new ArrayList<>(users)) {
            u.update(dt);
            maxWaitTime = Math.max(maxWaitTime, elapsedTime - u.spawnTimestamp);
        }

        users.removeIf(u -> u.removeMe);
        recalculateStats();
    }

    public void init() {
        for (ElevatorInterface ei : elevatorInterfaces) {
            ei.checkDestinationQueue();
        }
    }

    public void unwind() {
        for (Observable obj : new ArrayList<>(elevators)) obj.off("*");
        for (Observable obj : new ArrayList<>(elevatorInterfaces)) obj.off("*");
        for (Observable obj : new ArrayList<>(users)) obj.off("*");
        for (Observable obj : new ArrayList<>(floors)) obj.off("*");
        off("*");
        challengeEnded = true;
        elevators = new ArrayList<>();
        elevatorInterfaces = new ArrayList<>();
        users = new ArrayList<>();
        floors = new ArrayList<>();
    }
}

File: ports/python/elevatorsaga/world.py
"""World creation and simulation logic."""

import random
from .observable import Observable
from .floor import Floor
from .elevator import Elevator
from .user import User
from .interfaces import ElevatorInterface


class World(Observable):
    """The game world containing elevators, floors, and users."""

    def __init__(self, options=None):
        super().__init__()
        defaults = {
            "floor_height": 50,
            "floor_count": 4,
            "elevator_count": 2,
            "spawn_rate": 0.5,
            "elevator_capacities": [4],
        }
        if options:
            defaults.update(options)
        options = defaults

        self.floor_height = options["floor_height"]
        self.transported_counter = 0
        self.transported_per_sec = 0.0
        self.move_count = 0
        self.elapsed_time = 0.0
        self.max_wait_time = 0.0
        self.avg_wait_time = 0.0
        self.challenge_ended = False
        self.users = []
        self._spawn_rate = options["spawn_rate"]
        self._elapsed_since_spawn = 1.001 / self._spawn_rate

        def error_handler(e):
            self.trigger("usercode_error", e)

        # Create floors
        floor_count = options["floor_count"]
        self.floors = []
        for i in range(floor_count):
            y_pos = (floor_count - 1 - i) * self.floor_height
            self.floors.append(Floor(i, y_pos, error_handler))

        # Create elevators
        elevator_capacities = options["elevator_capacities"]
        current_x = 200.0
        self.elevators = []
        for i in range(options["elevator_count"]):
            cap = elevator_capacities[i % len(elevator_capacities)]
            elevator = Elevator(2.6, floor_count, self.floor_height, cap)
            elevator.move_to(current_x, None)
            elevator.set_floor_position(0)
            elevator.update_display_position()
            current_x += 20 + elevator.width
            self.elevators.append(elevator)

        # Create interfaces
        self.elevator_interfaces = [
            ElevatorInterface(e, floor_count, error_handler) for e in self.elevators
        ]

        # Bind elevator entrance events
        for elevator in self.elevators:
            elevator.on("entrance_available", self._handle_elev_availability)

        # Bind floor button repressing
        for floor in self.floors:
            floor.on("up_button_pressed", lambda f, ev="up_button_pressed": self._handle_button_repressing(ev, f))
            floor.on("down_button_pressed", lambda f, ev="down_button_pressed": self._handle_button_repressing(ev, f))

    def _handle_elev_availability(self, elevator):
        """Notify floors and users when elevator is available."""
        for i, floor in enumerate(self.floors):
            if elevator.current_floor == i:
                floor.elevator_available(elevator)
        for user in list(self.users):
            if user.current_floor == elevator.current_floor:
                user.elevator_available(elevator, self.floors[elevator.current_floor])

    def _handle_button_repressing(self, event_name, floor):
        """Handle button re-pressing to make elevators re-arrive."""
        offset = random.randint(0, len(self.elevators) - 1)
        for i in range(len(self.elevators)):
            elev_index = (i + offset) % len(self.elevators)
            elevator = self.elevators[elev_index]
            if (event_name == "up_button_pressed" and elevator.going_up_indicator) or \
               (event_name == "down_button_pressed" and elevator.going_down_indicator):
                if (elevator.current_floor == floor.level and
                        elevator.is_on_a_floor() and
                        not elevator.is_moving and
                        not elevator.is_full()):
                    self.elevator_interfaces[elev_index].go_to_floor(floor.level, True)
                    return

    def _recalculate_stats(self):
        """Recalculate world statistics."""
        if self.elapsed_time > 0:
            self.transported_per_sec = self.transported_counter / self.elapsed_time
        self.move_count = sum(e.move_count for e in self.elevators)
        self.trigger("stats_changed")

    def _register_user(self, user):
        """Register a new user in the world."""
        self.users.append(user)
        user.update_display_position(True)
        user.spawn_timestamp = self.elapsed_time
        self.trigger("new_user", user)

        def on_exited(elevator):
            self.transported_counter += 1
            self.max_wait_time = max(self.max_wait_time, self.elapsed_time - user.spawn_timestamp)
            self.avg_wait_time = (
                self.avg_wait_time * (self.transported_counter - 1) +
                (self.elapsed_time - user.spawn_timestamp)
            ) / self.transported_counter
            self._recalculate_stats()

        user.on("exited_elevator", on_exited)
        user.update_display_position(True)

    def _spawn_user_randomly(self):
        """Create and spawn a random user."""
        weight = random.randint(55, 100)
        user = User(weight)
        r = random.randint(0, 40)
        if r == 0:
            user.display_type = "child"
        elif random.randint(0, 1) == 0:
            user.display_type = "female"
        else:
            user.display_type = "male"

        user.move_to(105 + random.randint(0, 40), 0)
        floor_count = len(self.floors)
        current_floor = 0 if random.randint(0, 1) == 0 else random.randint(0, floor_count - 1)

        if current_floor == 0:
            destination_floor = random.randint(1, floor_count - 1)
        else:
            if random.randint(0, 10) == 0:
                destination_floor = (current_floor + random.randint(1, floor_count - 1)) % floor_count
            else:
                destination_floor = 0

        user.appear_on_floor(self.floors[current_floor], destination_floor)
        return user

    def update(self, dt):
        """Main world update for one time step."""
        self.elapsed_time += dt
        self._elapsed_since_spawn += dt

        while self._elapsed_since_spawn > 1.0 / self._spawn_rate:
            self._elapsed_since_spawn -= 1.0 / self._spawn_rate
            self._register_user(self._spawn_user_randomly())

        for elevator in self.elevators:
            elevator.update(dt)
            elevator.update_elevator_movement(dt)

        for user in self.users:
            user.update(dt)
            self.max_wait_time = max(self.max_wait_time, self.elapsed_time - user.spawn_timestamp)

        self.users = [u for u in self.users if not u.remove_me]
        self._recalculate_stats()

    def init(self):
        """Initialize the world - triggers idle events."""
        for interface in self.elevator_interfaces:
            interface.check_destination_queue()

    def unwind(self):
        """Clean up world resources."""
        for obj in self.elevators + self.elevator_interfaces + self.users + self.floors + [self]:
            obj.off("*")
        self.challenge_ended = True
        self.elevators = []
        self.elevator_interfaces = []
        self.users = []
        self.floors = []
